src/python/pants/backend/docker/goals/package_image_test.py

```diff
@@ -111,6 +113,11 @@ def assert_build(
     plugin_tags: tuple[str, ...] = (),
     expected_registries_metadata: None | list = None,
 ) -> None:
+    class DockerImageTagsRequestPlugin(DockerImageTagsRequest):
+        @classmethod
+        def image_tags_rule(cls):
```

Comment:
This is the meat of it. Basically, instead of `UnionRule` and all that jazz we lean in to subclassing. Unfortunately the engine won't currently accept a class member as an @rule (but that wouldn't be too hard to do once we're entirely call-by-name, and might not even be prohibitive now, it's just that `self` can't be type-annotated, and the rule mechanism doesn't like that, but we might be able to special-case it).

However right now we can use subclass polymorphism to return a by-name reference to the right rule to call.